repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py

```diff
@@ -0,0 +1,339 @@
+
+      returns string
+      """
+      removal_regexp = directive + '\s+' + value + '\s*\n'
+      removal_pattern = compile(removal_regexp)
+      content = sub(removal_pattern, '', content)
+
+      return content
+
+    def migrate_serverkeychain_path(self, path, new_directive, new_content):
+      """
+      migrate_serverkeychain_path
+
+      parameter:
+      - path - string
+      - new_directive - string
+      - new_content - string
+
+      returns string
+      """
+
+      # path can be relative to CUPS server root, our supported default is
+      # /etc/cups, so put it before it - that's how CUPS solves it in the code
+      if not path.startswith('/'):
+        path = '/etc/cups' + path
+      # move certificate or key to supported default path - /etc/cups/ssl - if it is not
+      # already
+      if not path.startswith('/etc/cups/ssl'):
+        os.rename(path, '/etc/cups/ssl/' + path.rsplit('/', 1)[1])
+        # if file was moved to default directory, no need to keep directive 
+        # in configuration file
+        new_content = self.remove_string(new_directive, path, new_content)
+      else:
+        if os.path.isfile(path):
+          # if path starts with /etc/cups/ssl, remove the filename from the path,
+          # because ServerKeychain wants directory
+          new_path = path.rsplit('/', 1)[0]
+          new_content = new_content.replace(path, new_path)
+
+
+      return new_content
+
+    def find_values(self, directive, content):
+      """
+      migrate_serverkeychain
+
+      parameters:
+      - directive - string
+      - new_content - string
+
+      returns list of strings
+      """
+
+      # we need to get values
+      value_regexp = '^\s*' + directive + '\s+(.+?)\s' 
+      value_pattern = compile(value_regexp, MULTILINE)
+      found_values = findall(value_pattern, content)
+
+      return found_values
+
+    def find_and_replace(self, directives, new_directive, file):
+      """
+      find_and_replace
+  
+      parameters:
+      - directives - list of strings
+      - new-directive - string
+      - file - string
+  
+      returns boolean
+      """
+
+      result = False
+
+      for directive in directives:
+        # regexp and new string differs if we replace directive or value - directive needs to have at least one
+        # white character at the end, but value needs to have one white character at start and zero or N white
+        # characters at the end + new line character
+        if directive in changed_directives:
+          directive_regexp = directive + '\s'
+          new_directive_string = new_directive + ' '
+        elif directive in changed_values:
+          directive_regexp = '\s' + directive + '\s*\n'
+          new_directive_string = ' ' + new_directive + '\n'
+        pattern = compile(directive_regexp)
+
+        try:
+          with open(file, 'r') as conf:
+            content = conf.read()
+        except IOError:
+          return result
```

Comment:
Would be interesting to return the error attached to the IOError exception. This can help us or users to understand better why an error happened.

Reply:
Done.